File: backend/app/api/v1/endpoints/tenants.py
"""
Tenant Management Endpoints for Phase 6A
Super Admin operations for managing tenants
"""
from fastapi import APIRouter, Depends, Query, Path, Request
from sqlalchemy.orm import Session
from typing import Optional, List
from uuid import UUID

from app.core.database import get_db
from app.api.deps import get_super_admin_user
from app.models.user import User
from app.models.branch import Branch
from app.services.tenant_service import TenantService
from app.schemas.tenant import (
    TenantCreate, TenantUpdate, TenantSubscriptionUpdate,
    TenantFeatureUpdate, TenantStatusUpdate, TenantResponse,
    TenantSummary, TenantStats, TenantListResponse, SystemStats
)
from app.schemas.branch import BranchResponse, BranchListResponse
from app.schemas.user import UserWithBranch, UserListResponse

router = APIRouter(prefix="/admin/tenants", tags=["Tenants"])


# ============================================================================
# COLLECTION ROUTES (no path params) - Must come FIRST
# ============================================================================

@router.post("/", response_model=TenantResponse, status_code=201)
def create_tenant(
    tenant_data: TenantCreate,
    request: Request,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_super_admin_user)
):
    """
    Create new tenant with admin user and HQ branch

    **Super Admin Only**

    This endpoint:
    - Creates a new tenant organization
    - Creates HQ branch automatically
    - Creates admin user for the tenant
    - Sets up default limits based on tier
    """
    service = TenantService(db)
    tenant = service.create_tenant(tenant_data, current_user, request)
    return tenant


@router.get("/", response_model=TenantListResponse)
def get_tenants(
    skip: int = Query(0, ge=0, description="Number of records to skip"),
    limit: int = Query(100, ge=1, le=500, description="Number of records to return"),
    tier: Optional[str] = Query(None, description="Filter by tier"),
    status: Optional[str] = Query(None, description="Filter by subscription status"),
    is_active: Optional[bool] = Query(None, description="Filter by active status"),
    search: Optional[str] = Query(None, description="Search by name or subdomain"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_super_admin_user)
):
    """
    Get all tenants with filters and pagination

    **Super Admin Only**

    Filters:
    - tier: free, basic, premium, enterprise
    - status: active, trial, expired, cancelled, suspended
    - is_active: true/false
    - search: searches in name and subdomain
    """
    service = TenantService(db)
    tenants, total = service.get_all_tenants(
        skip=skip,
        limit=limit,
        tier=tier,
        status=status,
        is_active=is_active,
        search=search
    )

    # Convert to summaries with user/branch counts
    summaries = []
    for tenant in tenants:
        summary = TenantSummary(
            id=tenant.id,
            name=tenant.name,
            subdomain=tenant.subdomain,
            tier=tenant.tier,
            subscription_status=tenant.subscription_status,
            is_active=tenant.is_active,
            created_at=tenant.created_at,
            user_count=len([u for u in tenant.users if u.is_active]),
            branch_count=len([b for b in tenant.branches if b.is_active])
        )
        summaries.append(summary)

    # Calculate pagination
    total_pages = (total + limit - 1) // limit
    page = (skip // limit) + 1

    return TenantListResponse(
        items=summaries,
        total=total,
        page=page,
        page_size=limit,
        total_pages=total_pages
    )


@router.get("/stats", response_model=SystemStats)
def get_system_stats(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_super_admin_user)
):
    """
    Get system-wide statistics

    **Super Admin Only**

    Returns:
    - Total tenants by status and tier
    - Total users and branches
    - Recent activity metrics
    - Expiring trials/subscriptions
    """
    service = TenantService(db)
    return service.get_system_stats()


@router.get("/subdomain/{subdomain}", response_model=TenantResponse)
def get_tenant_by_subdomain(
    subdomain: str = Path(..., description="Tenant subdomain"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_super_admin_user)
):
    """
    Get tenant by subdomain

    **Super Admin Only**

    Useful for looking up tenants by their subdomain identifier
    """
    service = TenantService(db)
    tenant = service.get_tenant_by_subdomain(subdomain)
    if not tenant:
        from app.core.exceptions import TenantNotFoundException
        raise TenantNotFoundException(f"Tenant with subdomain '{subdomain}' not found")
    return tenant


# ============================================================================
# NESTED ROUTES (/{tenant_id}/...) - Must come BEFORE single resource route
# ============================================================================

@router.get("/{tenant_id}/users", response_model=UserListResponse)
def get_tenant_users(
    tenant_id: UUID = Path(..., description="Tenant ID"),
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    search: Optional[str] = Query(None, description="Search by name or email"),
    role: Optional[str] = Query(None, description="Filter by role"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_super_admin_user)
):
    """
    Get all users for a specific tenant

    **Super Admin Only**

    Parameters:
    - tenant_id: The tenant's UUID
    - skip: Number of records to skip (pagination)
    - limit: Maximum records to return
    - search: Search by name or email
    - role: Filter by user role
    """
    from sqlalchemy import or_

    # Verify tenant exists
    service = TenantService(db)
    tenant = service.get_tenant_by_id(tenant_id)

    # Build query for this tenant's users
    query = db.query(User).filter(
        User.tenant_id == tenant_id,
        User.is_active == True
    )

    # Apply search filter
    if search:
        search_filter = or_(
            User.email.ilike(f"%{search}%"),
            User.full_name.ilike(f"%{search}%")
        )
        query = query.filter(search_filter)

    # Apply role filter
    if role:
        query = query.filter(User.role == role)

    # Get total count
    total = query.count()

    # Get paginated results
    users = query.offset(skip).limit(limit).all()

    # Convert to response with branch info
    users_with_info = []
    for user in users:
        user_dict = UserWithBranch.model_validate(user).model_dump()
        if user.default_branch:
            user_dict['branch_name'] = user.default_branch.name
            user_dict['branch_code'] = user.default_branch.code
        user_dict['tenant_name'] = tenant.name
        user_dict['tenant_subdomain'] = tenant.subdomain
        users_with_info.append(UserWithBranch(**user_dict))

    return UserListResponse(
        users=users_with_info,
        total=total,
        page=skip // limit + 1 if limit > 0 else 1,
        page_size=limit
    )


@router.get("/{tenant_id}/branches", response_model=BranchListResponse)
def get_tenant_branches(
    tenant_id: UUID = Path(..., description="Tenant ID"),
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_super_admin_user)
):
    """
    Get all branches for a specific tenant

    **Super Admin Only**
    """
    # Verify tenant exists
    service = TenantService(db)
    tenant = service.get_tenant_by_id(tenant_id)

    # Get branches for this tenant
    branches = db.query(Branch).filter(
        Branch.tenant_id == tenant_id,
        Branch.is_active == True
    ).offset(skip).limit(limit).all()

    total = db.query(Branch).filter(
        Branch.tenant_id == tenant_id,
        Branch.is_active == True
    ).count()

    return BranchListResponse(
        branches=[BranchResponse.model_validate(b) for b in branches],
        total=total,
        page=skip // limit + 1 if limit > 0 else 1,
        page_size=limit
    )


@router.get("/{tenant_id}/stats", response_model=TenantStats)
def get_tenant_statistics(
    tenant_id: UUID = Path(..., description="Tenant ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_super_admin_user)
):
    """
    Get detailed statistics for a specific tenant

    **Super Admin Only**

    Returns:
    - Usage vs limits (users, branches, storage)
    - Trial/expiry status
    - Last activity timestamp
    """
    service = TenantService(db)
    return service.get_tenant_stats(tenant_id)


@router.put("/{tenant_id}/subscription", response_model=TenantResponse)
def update_tenant_subscription(
    tenant_id: UUID = Path(..., description="Tenant ID"),
    subscription_data: TenantSubscriptionUpdate = ...,
    request: Request = ...,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_super_admin_user)
):
    """
    Update tenant subscription and limits

    **Super Admin Only**

    Can update:
    - tier (free, basic, premium, enterprise)
    - subscription_status
    - max_users, max_branches, max_storage_gb
    - trial_ends_at, subscription_ends_at
    """
    service = TenantService(db)
    return service.update_subscription(tenant_id, subscription_data, current_user, request)


@router.put("/{tenant_id}/features", response_model=TenantResponse)
def update_tenant_features(
    tenant_id: UUID = Path(..., description="Tenant ID"),
    feature_data: TenantFeatureUpdate = ...,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_super_admin_user)
):
    """
    Update tenant feature flags

    **Super Admin Only**

    Enable/disable specific features:
    - inventory_module
    - sales_module
    - pos_module
    - analytics
    - api_access
    - etc.
    """
    service = TenantService(db)
    return service.update_features(tenant_id, feature_data)


@router.patch("/{tenant_id}/status", response_model=TenantResponse)
def update_tenant_status(
    tenant_id: UUID = Path(..., description="Tenant ID"),
    status_data: TenantStatusUpdate = ...,
    request: Request = ...,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_super_admin_user)
):
    """
    Activate or deactivate tenant

    **Super Admin Only**

    - Sets is_active flag
    - Logs reason in metadata
    - When inactive, tenant users cannot login
    """
    service = TenantService(db)
    return service.update_status(tenant_id, status_data, current_user, request)


# ============================================================================
# SINGLE RESOURCE ROUTES (/{tenant_id}) - Must come LAST
# ============================================================================

@router.get("/{tenant_id}", response_model=TenantResponse)
def get_tenant(
    tenant_id: UUID = Path(..., description="Tenant ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_super_admin_user)
):
    """
    Get tenant details by ID

    **Super Admin Only**
    """
    service = TenantService(db)
    tenant = service.get_tenant_by_id(tenant_id)

    # Calculate user and branch counts
    user_count = len([u for u in tenant.users if u.is_active])
    branch_count = len([b for b in tenant.branches if b.is_active])

    # Return response with counts
    return TenantResponse(
        id=tenant.id,
        name=tenant.name,
        subdomain=tenant.subdomain,
        domain=tenant.domain,
        logo_url=tenant.logo_url,
        tier=tenant.tier,
        subscription_status=tenant.subscription_status,
        max_users=tenant.max_users,
        max_branches=tenant.max_branches,
        max_storage_gb=tenant.max_storage_gb,
        features=tenant.features or {},
        settings=tenant.settings or {},
        meta_data=tenant.meta_data or {},
        trial_ends_at=tenant.trial_ends_at,
        subscription_ends_at=tenant.subscription_ends_at,
        is_active=tenant.is_active,
        created_at=tenant.created_at,
        updated_at=tenant.updated_at,
        user_count=user_count,
        branch_count=branch_count
    )


@router.put("/{tenant_id}", response_model=TenantResponse)
def update_tenant(
    tenant_id: UUID = Path(..., description="Tenant ID"),
    update_data: TenantUpdate = ...,
    request: Request = ...,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_super_admin_user)
):
    """
    Update tenant basic information

    **Super Admin Only**

    Can update:
    - name
    - domain
    - logo_url
    - settings
    - meta_data
    """
    service = TenantService(db)
    return service.update_tenant(tenant_id, update_data, current_user, request)


@router.delete("/{tenant_id}", status_code=204)
def delete_tenant(
    tenant_id: UUID = Path(..., description="Tenant ID"),
    request: Request = ...,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_super_admin_user)
):
    """
    Delete (deactivate) tenant

    **Super Admin Only**

    This performs a soft delete:
    - Sets is_active = False
    - Sets deleted_at timestamp
    - Data is preserved for recovery

    For hard delete, use database operations directly
    """
    service = TenantService(db)
    service.delete_tenant(tenant_id, current_user, request)
    return None

File: frontend/src/lib/api/tenants.ts
/**
 * Tenant Management API Client
 * Phase 6B - Frontend Implementation
 * 
 * This module provides type-safe API methods for:
 * - Super Admin operations (tenant management)
 * - Tenant self-service operations
 */

import { apiClient } from './client';

// ============================================================================
// TYPE DEFINITIONS
// ============================================================================

export interface Tenant {
  id: string;
  name: string;
  subdomain: string;
  domain: string | null;
  tier: 'free' | 'basic' | 'premium' | 'enterprise';
  subscription_status: 'active' | 'trial' | 'expired' | 'cancelled' | 'suspended';
  max_users: number;
  max_branches: number;
  max_storage_gb: number;
  features: Record<string, boolean>;
  settings: Record<string, unknown>;
  meta_data: Record<string, unknown>;
  trial_ends_at: string | null;
  subscription_ends_at: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string | null;
  logo_url: string | null;
}

export interface TenantSummary {
  id: string;
  name: string;
  subdomain: string;
  tier: string;
  subscription_status: string;
  is_active: boolean;
  created_at: string;
  user_count: number;
  branch_count: number;
}

export interface TenantStats {
  id: string;
  name: string;
  subdomain: string;
  tier: string;
  subscription_status: string;
  max_users: number;
  max_branches: number;
  max_storage_gb: number;
  user_count: number;
  branch_count: number;
  storage_used_gb: number;
  users_usage_percent: number;
  branches_usage_percent: number;
  storage_usage_percent: number;
  is_active: boolean;
  is_trial: boolean;
  is_expired: boolean;
  days_until_expiry: number | null;
  created_at: string;
  last_activity_at: string | null;
}

export interface SystemStats {
  total_tenants: number;
  active_tenants: number;
  inactive_tenants: number;
  trial_tenants: number;
  free_tier_count: number;
  basic_tier_count: number;
  premium_tier_count: number;
  enterprise_tier_count: number;
  total_users: number;
  total_branches: number;
  tenants_created_today: number;
  tenants_created_this_week: number;
  tenants_created_this_month: number;
  trials_expiring_soon: number;
  subscriptions_expiring_soon: number;
}

export interface TenantUsage {
  tenant_id: string;
  tenant_name: string;
  tier: string;
  users_current: number;
  users_limit: number;
  users_available: number;
  users_percent: number;
  branches_current: number;
  branches_limit: number;
  branches_available: number;
  branches_percent: number;
  storage_used_gb: number;
  storage_limit_gb: number;
  storage_available_gb: number;
  storage_percent: number;
  is_user_limit_reached: boolean;
  is_branch_limit_reached: boolean;
  is_storage_limit_reached: boolean;
  can_upgrade: boolean;
  next_tier: string | null;
}

export interface TierInfo {
  tier: string;
  display_name: string;
  price_monthly: number;
  price_yearly: number;
  max_users: number;
  max_branches: number;
  max_storage_gb: number;
  features: string[];
  is_recommended: boolean;
}

export interface CreateTenantData {
  name: string;
  subdomain: string;
  domain?: string;
  tier: string;
  admin_email: string;
  admin_password: string;
  admin_first_name: string;
  admin_last_name: string;
  max_users?: number;
  max_branches?: number;
  max_storage_gb?: number;
  logo_url?: string;
}

export interface UpdateTenantData {
  name?: string;
  domain?: string;
  logo_url?: string;
  settings?: Record<string, unknown>;
  meta_data?: Record<string, unknown>;
}

export interface UpdateSubscriptionData {
  tier: string;
  subscription_status: string;
  max_users: number;
  max_branches: number;
  max_storage_gb: number;
  trial_ends_at?: string;
  subscription_ends_at?: string;
}

export interface UpdateFeaturesData {
  features: Record<string, boolean>;
}

export interface UpdateStatusData {
  is_active: boolean;
  reason?: string;
}

export interface TenantListParams {
  skip?: number;
  limit?: number;
  tier?: string;
  status?: string;
  is_active?: boolean;
  search?: string;
}

export interface TenantListResponse {
  items: TenantSummary[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// ============================================================================
// API CLIENT
// ============================================================================

export const tenantsAPI = {
  // ========================================================================
  // SUPER ADMIN OPERATIONS
  // ========================================================================

  /**
   * Get system-wide statistics (Super Admin only)
   */
  async getSystemStats(): Promise<SystemStats> {
    const response = await apiClient.get<SystemStats>('/admin/tenants/stats');
    return response;
  },

  /**
   * List all tenants with filters and pagination (Super Admin only)
   */
  async listTenants(params?: TenantListParams): Promise<TenantListResponse> {
    const response = await apiClient.get<TenantListResponse>('/admin/tenants/', {
      params,
    });
    return response;
  },

  /**
   * Get tenant details by ID (Super Admin only)
   */
  async getTenant(tenantId: string): Promise<Tenant> {
    const response = await apiClient.get<Tenant>(`/admin/tenants/${tenantId}`);
    return response;
  },

  /**
   * Get tenant statistics by ID (Super Admin only)
   */
  async getTenantStats(tenantId: string): Promise<TenantStats> {
    const response = await apiClient.get<TenantStats>(
      `/admin/tenants/${tenantId}/stats`
    );
    return response;
  },

  /**
   * Get tenant by subdomain (Super Admin only)
   */
  async getTenantBySubdomain(subdomain: string): Promise<Tenant> {
    const response = await apiClient.get<Tenant>(
      `/admin/tenants/subdomain/${subdomain}`
    );
    return response;
  },

  /**
   * Create new tenant (Super Admin only)
   */
  async createTenant(data: CreateTenantData): Promise<Tenant> {
    const response = await apiClient.post<Tenant>('/admin/tenants/', data);
    return response;
  },

  /**
   * Update tenant basic information (Super Admin only)
   */
  async updateTenant(
    tenantId: string,
    data: UpdateTenantData
  ): Promise<Tenant> {
    const response = await apiClient.put<Tenant>(
      `/admin/tenants/${tenantId}`,
      data
    );
    return response;
  },

  /**
   * Update tenant subscription (Super Admin only)
   */
  async updateSubscription(
    tenantId: string,
    data: UpdateSubscriptionData
  ): Promise<Tenant> {
    const response = await apiClient.put<Tenant>(
      `/admin/tenants/${tenantId}/subscription`,
      data
    );
    return response;
  },

  /**
   * Update tenant features (Super Admin only)
   */
  async updateFeatures(
    tenantId: string,
    data: UpdateFeaturesData
  ): Promise<Tenant> {
    const response = await apiClient.put<Tenant>(
      `/admin/tenants/${tenantId}/features`,
      data
    );
    return response;
  },

  /**
   * Update tenant status (activate/deactivate) (Super Admin only)
   */
  async updateStatus(
    tenantId: string,
    data: UpdateStatusData
  ): Promise<Tenant> {
    const response = await apiClient.patch<Tenant>(
      `/admin/tenants/${tenantId}/status`,
      data
    );
    return response;
  },

  /**
   * Delete tenant (Super Admin only)
   */
  async deleteTenant(tenantId: string): Promise<void> {
    await apiClient.delete(`/admin/tenants/${tenantId}`);
  },

  // ========================================================================
  // TENANT SELF-SERVICE OPERATIONS
  // ========================================================================

  /**
   * Get current tenant information
   */
  async getMyTenant(): Promise<Tenant> {
    const response = await apiClient.get<Tenant>('/tenant/');
    return response;
  },

  /**
   * Get current tenant usage statistics
   */
  async getUsage(): Promise<TenantUsage> {
    const response = await apiClient.get<TenantUsage>('/tenant/usage');
    return response;
  },

  /**
   * Get available subscription tiers
   */
  async getTiers(): Promise<{ tiers: TierInfo[]; current_tier: string }> {
    const response = await apiClient.get<{
      tiers: TierInfo[];
      current_tier: string;
    }>('/tenant/tiers');
    return response;
  },

  /**
   * Check user limit
   */
  async checkUserLimit(): Promise<{
    can_add: boolean;
    current_count: number;
    limit: number;
    available: number;
    percentage: number;
  }> {
    const response = await apiClient.get<{
        can_add: boolean;
        current_count: number;
        limit: number;
        available: number;
        percentage: number;
    }>('/tenant/limits/users');
    return response;
  },

  /**
   * Check branch limit
   */
  async checkBranchLimit(): Promise<{
    can_add: boolean;
    current_count: number;
    limit: number;
    available: number;
    percentage: number;
  }> {
    const response = await apiClient.get<{
        can_add: boolean;
        current_count: number;
        limit: number;
        available: number;
        percentage: number;
    }>('/tenant/limits/branches');
    return response;
  },

  /**
   * Check feature access
   */
  async checkFeature(featureName: string): Promise<{
    feature_name: string;
    enabled: boolean;
  }> {
    const response = await apiClient.get<{
        feature_name: string;
        enabled: boolean;
    }>(`/tenant/features/${featureName}`);
    return response;
  },

  /**
   * Update tenant settings (self-service)
   */
  async updateSettings(data: {
    name?: string;
    logo_url?: string;
    settings?: Record<string, unknown>;
  }): Promise<Tenant> {
    const response = await apiClient.put<Tenant>('/tenant/settings', data);
    return response;
  },
};

// ============================================================================
// HELPER FUNCTIONS
// ============================================================================

/**
 * Get tier color class for badges
 */
export function getTierColor(tier: string): string {
  switch (tier) {
    case 'free':
      return 'text-gray-700 dark:text-gray-200 bg-gray-100 dark:bg-gray-800';
    case 'basic':
      return 'text-blue-700 dark:text-blue-200 bg-blue-100 dark:bg-blue-900/50';
    case 'premium':
      return 'text-purple-700 dark:text-purple-200 bg-purple-100 dark:bg-purple-900/50';
    case 'enterprise':
      return 'text-amber-700 dark:text-amber-200 bg-amber-100 dark:bg-amber-900/50';
    default:
      return 'text-gray-700 dark:text-gray-200 bg-gray-100 dark:bg-gray-800';
  }
}

/**
 * Get status color class for badges
 */
export function getStatusColor(status: string): string {
  switch (status) {
    case 'active':
      return 'text-green-700 dark:text-green-200 bg-green-100 dark:bg-green-900/50';
    case 'trial':
      return 'text-blue-700 dark:text-blue-200 bg-blue-100 dark:bg-blue-900/50';
    case 'expired':
      return 'text-red-700 dark:text-red-200 bg-red-100 dark:bg-red-900/50';
    case 'cancelled':
      return 'text-gray-700 dark:text-gray-200 bg-gray-100 dark:bg-gray-800';
    case 'suspended':
      return 'text-orange-700 dark:text-orange-200 bg-orange-100 dark:bg-orange-900/50';
    default:
      return 'text-gray-700 dark:text-gray-200 bg-gray-100 dark:bg-gray-800';
  }
}

/**
 * Get usage color based on percentage
 */
export function getUsageColor(percentage: number): string {
  if (percentage >= 90) return 'text-red-600';
  if (percentage >= 70) return 'text-yellow-600';
  return 'text-green-600';
}

/**
 * Format tier display name
 */
export function formatTierName(tier: string): string {
  return tier.charAt(0).toUpperCase() + tier.slice(1);
}

/**
 * Check if limit is reached
 */
export function isLimitReached(current: number, limit: number): boolean {
  return current >= limit;
}

/**
 * Calculate usage percentage
 */
export function calculateUsagePercent(current: number, limit: number): number {
  if (limit === 0) return 0;
  return Math.min((current / limit) * 100, 100);
}